### ISSUE TITLE
[Wotlk] Party Pets

### ISSUE THREAD

**Issue by XyzKangUI** [Verubato/frame-sort#4]
When party pets are enabled in wrath it will not sort because the GetUnits() function apparently does not account for partypet..i or raidpet..i units. So the amount of units will be less than the frames displayed and return on the LayoutRaid() function.

So I edited the GetUnits function to include pet units and it messes up the sorting creating a gap when you sort 'player' to be on the bottom:

What I've noticed are these issues:

1. When 'displayPets' is disabled (after editing GetUnits) it will detect 1 more unit than frames -> aborting the sorting. I didnt add a check for the raidOptionDisplayPets cvar here :/
2. When 'displayPets' is enabled (after the edit again) it will sort, but when you sort yourself on the 'bottom' it creates a gap as seen in the picture below (1a)
3. When 'displayPets' is enabled (idem) you cant sort yourself to the 'middle'. The funny thing is that when you do enable 'middle' it will sort you on the bottom and the pet frame will be attached to you, resulting in the correct behaviour for 'bottom' option. See picture below (1b)

1a. 
![image](https://user-images.githubusercontent.com/94811434/221423889-1708a405-08f7-42f5-bb8a-96066e01a16b.png)

1b.
![image](https://user-images.githubusercontent.com/94811434/221424650-c439633b-65f3-4542-a16c-c735ec3cb082.png)

**Comment by Verubato** [comment]
Ah right, I've totally missed party pets as they aren't working in retail at all (so people are using addons/weak auras).

I'll start working on fixing this.

**Comment by Verubato** [comment]
I've just pushed a potential fix for this, but I have limited ability to test this in Wotlk.
Reckon you could git clone the latest changes down and see if it's working?

**Comment by Verubato** [comment]
If by chance you have already done this, I've just re-released it as it had a bug, so please update again.

**Comment by XyzKangUI** [comment]
It broke sorting:

```
7x FrameSort/Core/Core.lua:186: attempt to index field 'frame' (a nil value)
[string "@FrameSort/Core/Core.lua"]:186: in function `ShuffleFrames'
[string "@FrameSort/Core/Core.lua"]:148: in function `LayoutRaid'
[string "@FrameSort/Loader.lua"]:35: in function <FrameSort/Loader.lua:35>
[string "=[C]"]: in function `CompactRaidFrameContainer_LayoutFrames'
[string "@Blizzard_CompactRaidFrames/Blizzard_CompactRaidFrameContainer.lua"]:130: in function `CompactRaidFrameContainer_TryUpdate'
[string "@Blizzard_CompactRaidFrames/Blizzard_CompactRaidFrameContainer.lua"]:57: in function `CompactRaidFrameContainer_OnEvent'
[string "*Blizzard_CompactRaidFrameContainer.xml:79_OnEvent"]:1: in function <[string "*Blizzard_CompactRaidFrameContainer.xml:79_..."]:1>

Locals:
self = <table> {
 TrySort = <function> defined @FrameSort/Core/Core.lua:46
 TrySortTraditional = <function> defined @FrameSort/Core/Core.lua:64
 SetTargets = <function> defined @FrameSort/Core/Target.lua:28
 Loader = Frame {
 }
 ShuffleFrames = <function> defined @FrameSort/Core/Core.lua:177
 SortPending = true
 GetRaidFrames = <function> defined @FrameSort/Util/Frame.lua:6
 OptionsBuilder = <table> {
 }
 TrySortTaintless = <function> defined @FrameSort/Core/Core.lua:93
 InitOptions = <function> defined @FrameSort/Config/Options.lua:259
 CompareTopLeft = <function> defined @FrameSort/Core/Compare.lua:116
 GetSortMode = <function> defined @FrameSort/Core/Compare.lua:33
 ToFrameWithPosition = <function> defined @FrameSort/Type/FrameWithPosition.lua:10
 CompareMiddle = <function> defined @FrameSort/Core/Compare.lua:88
 UpdateTargets = <function> defined @FrameSort/Core/Target.lua:16
 SortMode = <table> {
 }
 GetPartyFrames = <function> defined @FrameSort/Util/Frame.lua:34
 InitTargeting = <function> defined @FrameSort/Core/Target.lua:6
 SetOption = <function> defined @FrameSort/Config/Options.lua:249
 EventLoop = Frame {
 }
 GetPets = <function> defined @FrameSort/Util/Unit.lua:45
 GetSortFunction = <function> defined @FrameSort/Core/Compare.lua:7
 Init = <function> defined @FrameSort/Loader.lua:13
 OnLoadAddon = <function> defined @FrameSort/Loader.lua:5
 IsMember = <function> defined @FrameSort/Util/Unit.lua:73
 Options = <table> {
 }
 OnEvent = <function> defined @FrameSort/Core/Core.lua:5
 IsPet = <function> defined @FrameSort/Util/Unit.lua:60
 Defaults = <table> {
 }
 Compare = <function> defined @FrameSort/Core/Compare.lua:54
 Debug = <function> defined @FrameSort/Util/Log.lua:6
 UpgradeOptions = <function> defined @FrameSort/Config/Options.lua:197
 LayoutRaid = <function> defined @FrameSort/Core/Core.lua:113
 CanSort = <function> defined @FrameSort/Core/Core.lua:14
 GetUnits = <function> defined @FrameSort/Util/Unit.lua:21
}
orderedUnits = <table> {
 1 = "party1"
 2 = "player"
}
framesByUnit = <table> {
 party1 = <table> {
 }
 player = <table> {
 }
}
framesWithPoints = <table> {
 1 = <table> {
 }
 2 = <table> {
 }
}
(for index) = 1
(for limit) = 2
(for step) = 1
i = 1
sourceUnit = "party1"
source = <table> {
 Points = <table> {
 }
 Frame = CompactRaidFrame2 {
 }
}
target = <table> {
 Points = <table> {
 }
 Frame = CompactRaidFrame1 {
 }
}
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = "attempt to index field 'frame' (a nil value)"
```

**Comment by Verubato** [comment]
Whoops, ok just committed the fix for that.

**Comment by XyzKangUI** [comment]
```
6x FrameSort/Core/Core.lua:186: attempt to index local 'source' (a nil value)
[string "@FrameSort/Core/Core.lua"]:186: in function `ShuffleFrames'
[string "@FrameSort/Core/Core.lua"]:168: in function `LayoutRaid'
[string "@FrameSort/Loader.lua"]:35: in function <FrameSort/Loader.lua:35>
[string "=[C]"]: in function `CompactRaidFrameContainer_LayoutFrames'
[string "@Blizzard_CompactRaidFrames/Blizzard_CompactRaidFrameContainer.lua"]:130: in function `CompactRaidFrameContainer_TryUpdate'
[string "@Blizzard_CompactRaidFrames/Blizzard_CompactRaidFrameContainer.lua"]:57: in function `CompactRaidFrameContainer_OnEvent'
[string "*Blizzard_CompactRaidFrameContainer.xml:79_OnEvent"]:1: in function <[string "*Blizzard_CompactRaidFrameContainer.xml:79_..."]:1>

Locals:
self = <table> {
 TrySort = <function> defined @FrameSort/Core/Core.lua:46
 TrySortTraditional = <function> defined @FrameSort/Core/Core.lua:64
 SetTargets = <function> defined @FrameSort/Core/Target.lua:28
 Loader = Frame {
 }
 ShuffleFrames = <function> defined @FrameSort/Core/Core.lua:177
 GetRaidFrames = <function> defined @FrameSort/Util/Frame.lua:6
 OptionsBuilder = <table> {
 }
 TrySortTaintless = <function> defined @FrameSort/Core/Core.lua:93
 InitOptions = <function> defined @FrameSort/Config/Options.lua:259
 CompareTopLeft = <function> defined @FrameSort/Core/Compare.lua:116
 GetSortMode = <function> defined @FrameSort/Core/Compare.lua:33
 ToFrameWithPosition = <function> defined @FrameSort/Type/FrameWithPosition.lua:10
 CompareMiddle = <function> defined @FrameSort/Core/Compare.lua:88
 UpdateTargets = <function> defined @FrameSort/Core/Target.lua:16
 SortMode = <table> {
 }
 GetPartyFrames = <function> defined @FrameSort/Util/Frame.lua:34
 InitTargeting = <function> defined @FrameSort/Core/Target.lua:6
 SetOption = <function> defined @FrameSort/Config/Options.lua:249
 EventLoop = Frame {
 }
 GetPets = <function> defined @FrameSort/Util/Unit.lua:45
 GetSortFunction = <function> defined @FrameSort/Core/Compare.lua:7
 Init = <function> defined @FrameSort/Loader.lua:13
 OnLoadAddon = <function> defined @FrameSort/Loader.lua:5
 IsMember = <function> defined @FrameSort/Util/Unit.lua:73
 Options = <table> {
 }
 OnEvent = <function> defined @FrameSort/Core/Core.lua:5
 IsPet = <function> defined @FrameSort/Util/Unit.lua:60
 Defaults = <table> {
 }
 Compare = <function> defined @FrameSort/Core/Compare.lua:54
 Debug = <function> defined @FrameSort/Util/Log.lua:6
 UpgradeOptions = <function> defined @FrameSort/Config/Options.lua:197
 LayoutRaid = <function> defined @FrameSort/Core/Core.lua:113
 CanSort = <function> defined @FrameSort/Core/Core.lua:14
 GetUnits = <function> defined @FrameSort/Util/Unit.lua:21
}
orderedUnits = <table> {
 1 = "party1pet"
}
framesByUnit = <table> {
 partypet1 = <table> {
 }
}
framesWithPoints = <table> {
 1 = <table> {
 }
}
(for index) = 1
(for limit) = 1
(for step) = 1
i = 1
sourceUnit = "party1pet"
source = nil
target = <table> {
 points = <table> {
 }
 frame = CompactRaidFrame4 {
 }
}
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = nil
(*temporary) = "attempt to index local 'source' (a nil value)"
```

**Comment by XyzKangUI** [comment]
The error is caused by the pet unit, although you seem to have fixed top/middle/bottom sorting when there is a pet frame showing.

**Comment by Verubato** [comment]
Is the pet unit called party1pet or partypet1?

**Comment by XyzKangUI** [comment]
> Is the pet unit called party1pet or partypet1?

both work, i just commented out addon:ShuffleFrames(pets, petFramesByUnit, petFramesWithPoints) and it works fine

**Comment by Verubato** [comment]
Ok cool, so I think there is just 1 bug left I need to fix to get the pets sorted as well then.

**Comment by XyzKangUI** [comment]
> Ok cool, so I think there is just 1 bug left I need to fix to get the pets sorted as well then.

why do you want to sort the pets, they are always attached to the bottom frame. Seems to be the correct behaviour

**Comment by Verubato** [comment]
I was thinking for example if the sorted party is:

party2
party3
party1
player

Then the pets should be:

partypet2
partypet3
partypet1
playerpet

**Comment by XyzKangUI** [comment]
I like the commitment, but I dont think that the average player cares about pets position. It are mostly the PvP'ers who like to sort their frames by unit because of their macro's.

**Comment by Verubato** [comment]
For one of the party pet frames, would you mind doing a `/dump CompactRaidFrameXYZ.unit` where XYZ is the number of the pet frame?

You might know already, but you can find the frame name by doing /fstack and then mouseover the frames to find out it's name.

**Comment by Verubato** [comment]
> 

Yeah true, I think I'm so close to having it working so might as well try it anyway.

**Comment by XyzKangUI** [comment]
I believe that the unit names are just the default game units. From the lock PoV his pet will be "pet" and from the priest PoV the lock's pet will be "partypet1". When you convert the party to a raid it will be "raidpet1" for both the lock and priest, but macro's that use "pet" or "partypet1" still remain intact.

**Comment by XyzKangUI** [comment]
Although I guess that it answers your question if I tell you that its partypet..index and raidpet..index

**Comment by Verubato** [comment]
Yeah I'm guessing this last bug is because i'm doing `framesByUnit[partypet1]` but actually the frame unit is `framesByUnit[party1pet]` or something like that. Just trying to think of someway to fix this cleanly at the moment.

**Comment by XyzKangUI** [comment]
Here it confirms the statement made above:
"player":
![image](https://user-images.githubusercontent.com/94811434/221640943-4e3afc23-6150-451e-bbd9-01045bfa3e19.png)

from non player pov: 
![image](https://user-images.githubusercontent.com/94811434/221641118-513a8ea8-02ca-4e15-a06d-388052c17839.png)

**Comment by XyzKangUI** [comment]
You can use the SecureButton_GetUnit API to get an unit from a raidframe, e.g. SecureButton_GetUnit(CompactRaidFrame7) will return the unit id attached to it.

**Comment by Verubato** [comment]
I've just pushed a potential fix if you wouldn't mind doing another check? Thanks so much for your help with this, I'm not sure how I could test this on Wotlk other than asking for randoms to join group.

**Comment by XyzKangUI** [comment]
I made like 4 accounts for wrath PTR servers so that I can do tests on. Pretty solid.

**Comment by Verubato** [comment]
Oh I see, nice I'll do the same!

**Comment by Verubato** [comment]
Oh ok, that sounds like a better approach, I will update to use SecureButton_GetUnit.

**Comment by XyzKangUI** [comment]
`units[#units + 1] = SecureButton_GetUnit(frame[i])`

This line will cause an error.

**Comment by Verubato** [comment]
Yeah just fixed that one a second ago (amended my commit).
Getting a new error now that I'm looking into.

**Comment by Verubato** [comment]
Alright I think I've got it working well enough for a release.
There might be some more refactoring I can do there which I'll investigate later, but it seems I can't quite get away from needing the aliases for pets.

**Comment by XyzKangUI** [comment]
I am curious if one day you will expand this addon's features to change spacing levels between frames. I am aware you are storing the X and Y values but if it could be used to add an own value on top , e.g. +5 it would be amazing.

**Comment by Verubato** [comment]
I was actually thinking of adding in that feature! It would be pretty easy to add in some +-offsets.
I will work on adding this over the next few days.
If you have any other ideas/suggestions in the future as well feel free to let me know :)

**Comment by Verubato** [comment]
Issue fixed in v3.2.0.
